service/history/hsm/tree.go

```diff
@@ -185,22 +203,56 @@ func (n *Node) Path() []Key {
 
 // Outputs returns all outputs produced by transitions on this tree.
 func (n *Node) Outputs() []PathAndOutputs {
+	root := n.getRoot()
+	if root.opLog == nil {
+		return nil
+	}
+
+	currentPath := n.Path()
 	var paos []PathAndOutputs
-	if len(n.cache.outputs) > 0 {
-		paos = append(paos, PathAndOutputs{Path: n.Path(), Outputs: n.cache.outputs})
+
+	isDeleted := func(path []Key) bool {
```

Comment:
nit: this can be a method on `OperationLog`.

Reply:
Done